src/tracking-buffer/writable-tracking-buffer.js

```diff
@@ -93,137 +102,140 @@ module.exports = class WritableTrackingBuffer {
     this.buffer.writeInt16BE(value, this.position);
     return this.position += length;
   }
 
-  writeInt32LE(value) {
+  writeInt32LE(value: number) {
     const length = 4;
     this.makeRoomFor(length);
     this.buffer.writeInt32LE(value, this.position);
     return this.position += length;
   }
 
-  writeInt32BE(value) {
+  writeInt32BE(value: number) {
     const length = 4;
     this.makeRoomFor(length);
     this.buffer.writeInt32BE(value, this.position);
     return this.position += length;
   }
 
-  writeFloatLE(value) {
+  writeFloatLE(value: number) {
     const length = 4;
     this.makeRoomFor(length);
     this.buffer.writeFloatLE(value, this.position);
     return this.position += length;
   }
 
-  writeDoubleLE(value) {
+  writeDoubleLE(value: number) {
     const length = 8;
     this.makeRoomFor(length);
     this.buffer.writeDoubleLE(value, this.position);
     return this.position += length;
   }
 
-  writeString(value, encoding) {
-    encoding || (encoding = this.encoding);
+  writeString(value: string, encoding: ?Encoding) {
+    if (encoding == undefined) {
+      encoding = this.encoding;
+    }
 
     const length = Buffer.byteLength(value, encoding);
     this.makeRoomFor(length);
 
-    const bytesWritten = this.buffer.write(value, this.position, encoding);
+    const bytesWritten = this.buffer.write(value, this.position, this.buffer.length - this.position, encoding);
     this.position += length;
 
     return bytesWritten;
   }
 
-  writeBVarchar(value, encoding) {
+  writeBVarchar(value: string, encoding: ?Encoding) {
     this.writeUInt8(value.length);
     return this.writeString(value, encoding);
   }
 
-  writeUsVarchar(value, encoding) {
+  writeUsVarchar(value: string, encoding: ?Encoding) {
     this.writeUInt16LE(value.length);
     return this.writeString(value, encoding);
   }
 
-  writeUsVarbyte(value, encoding) {
-    if (encoding == null) {
+  // TODO: Figure out what types are passed in other than `Buffer`
+  writeUsVarbyte(value: any, encoding: ?Encoding) {
+    if (encoding == undefined) {
       encoding = this.encoding;
     }
 
     let length;
-    if (Buffer.isBuffer(value)) {
+    if (value instanceof Buffer) {
       length = value.length;
     } else {
       value = value.toString();
       length = Buffer.byteLength(value, encoding);
     }
     this.writeUInt16LE(length);
 
-    if (Buffer.isBuffer(value)) {
+    if (value instanceof Buffer) {
       return this.writeBuffer(value);
     } else {
       this.makeRoomFor(length);
-      this.buffer.write(value, this.position, encoding);
+      this.buffer.write(value, this.position, this.buffer.length - this.position, encoding);
       return this.position += length;
     }
   }
 
-  writePLPBody(value, encoding) {
+  writePLPBody(value: any, encoding: ?Encoding) {
     if (encoding == null) {
       encoding = this.encoding;
     }
 
     let length;
-    if (Buffer.isBuffer(value)) {
+    if (value instanceof Buffer) {
```

Comment:
Flow does not understand that the `Buffer.isBuffer` check refines the type of the `value` variable. But switching to an `instanceof` check here does the trick.